mteb/cross_encoder_interface.py

```diff
@@ -0,0 +1,46 @@
+from __future__ import annotations
+
+from collections.abc import Sequence
+from typing import Any, Protocol, Union, runtime_checkable
+
+import numpy as np
+import torch
+
+Corpus = Union[list[dict[str, str]], dict[str, list[str]]]
+
+
+@runtime_checkable
+class CrossEncoder(Protocol):
+    """The interface for a cross-encoder in MTEB.
+
+    In general the interface is kept aligned with sentence-transformers interface. In cases where exceptions occurs these are handled within MTEB.
+    """
+
+    def __init__(self, device: str | None = None) -> None:
+        """The initialization function for the cross-encoder. Used when calling it from the mteb run CLI.
+
+        Args:
+            device: The device to use for prediction. Can be ignored if the encoder is not using a device (e.g. for API)
+        """
+
+    def predict(
+        self,
+        queries: Sequence[str],
+        passages: Sequence[str],
+        *,
+        task_name: str | None = None,
+        instruction: str | None = None,
+        **kwargs: Any,
+    ) -> np.ndarray | torch.Tensor:
+        """Predicts scores for query-passage pairs. Note that unlike the encoder, the prompt is specified in the model as it is typically more complex and model-specific.
```

Comment:
```suggestion
        """Predicts relevance scores for query-passage pairs. Note that, unlike the encoder, the cross-encoder can compare across queries and passages.
        """
```

> the prompt is specified in the model as it is typically more complex and model-specific.

Hmm, I am not sure I follow. You can implement arbitrarily complex prompts using just the task name.

Reply:
I may be misunderstanding the prompts then. Does the task name allow for slotting arguments?

e.g. 
prompt = """{query}\n\nSome text here: {passage}\n\n{instruction} closing text here"""? This is what most rerankers use something like.

Reply:
@Samoed if you know as well?

Reply:
I'm not sure what you mean, but you can specify instructions for tasks. However, I don't think this applies to templates.

Reply:
The model will get the task_name and from there they can essentially implement their own mapping from task to prompt/template.

Reply:
> Got it, so you're saying define a model_prompts which has the main prompt and then get it back via the task_name?

yes

> Is it okay if we leave it as is right now, and push this to a separate PR? Perhaps there are other use cases but I can't think of them currently.

Hmm would much rather add the complexity when needed. Removing instructions here would allow us to have the CrossEncoder and Encoder's encode interface be exactly the same (which would greatly simplify things)

Reply:
Okay, will remove the crossencoder stuff from this PR then and we can handle it in another.  That’s a much bigger change and I have others waiting on the Promptriever fix.

Reply:
Hmm, sorry you are right, leave it as is and get this fix merged in, we can then update it in an upcoming (easier that splitting it up)